internal/core/processor/universal/processor.go

```diff
@@ -0,0 +1,228 @@
+		return p.configStartingBlock, nil
+	}
+
+	startingBlock, err := p.client.BlockNumber(ctx)
+	if err != nil {
+		return 0, fmt.Errorf("error retrieving the latest block from chain: %w", err)
+	}
+
+	slog.Debug("using latestBlock from blockchain as startingBlock", "startingBlock", startingBlock)
+	return startingBlock, nil
+}
+
+func (p *processor) GetLastBlock(ctx context.Context, startingBlock uint64) (uint64, error) {
+	l1LatestBlock, err := p.client.BlockNumber(ctx)
+	if err != nil {
+		slog.Error("error retrieving the latest block", "err", err.Error())
+		return 0, err
+	}
+
+	return min(startingBlock+p.configBlocksRange, l1LatestBlock-p.configBlocksMargin), nil
+}
+
+func (p *processor) VerifyChainConsistency(ctx context.Context, startingBlock uint64) error {
+	tx := p.stateService.NewTransaction()
+	defer tx.Discard()
+
+	lastBlockDB, err := tx.GetLastOwnershipBlock()
+	if err != nil {
+		slog.Error("error occurred while reading LaosEvolution end range block hash", "err", err.Error())
+		return err
+	}
+
+	// During the initial iteration, no hash is stored in the database, so this code block is bypassed.
+	if (lastBlockDB.Hash == common.Hash{}) {
+		return nil
+	}
+
+	// Verify whether the hash of the last block from the previous iteration remains unchanged;
+	// if it differs, it indicates a reorganization has taken place.
+	previousLastBlock := startingBlock - 1
+	slog.Debug("verifying chain consistency on block number", "previousLastBlock", previousLastBlock)
+	previousLastBlockData, err := p.client.BlockByNumber(ctx, big.NewInt(int64(previousLastBlock)))
+	if err != nil {
+		slog.Error("error occurred while retrieving new start range block", "err", err.Error())
+		return err
+	}
+
+	if previousLastBlockData.Hash().Cmp(lastBlockDB.Hash) != 0 {
+		return ReorgError{Block: previousLastBlock, ChainHash: previousLastBlockData.Hash(), StorageHash: lastBlockDB.Hash}
+	}
+
+	return nil
+}
+
+func (p *processor) IsEvoSyncedWithOwnership(ctx context.Context, lastOwnershipBlock uint64) (bool, error) {
+	lastBlockHeader, err := p.client.HeaderByNumber(ctx, big.NewInt(int64(lastOwnershipBlock)))
+	if err != nil {
+		return false, err
+	}
+
+	tx := p.stateService.NewTransaction()
+	defer tx.Discard()
+
+	lastEvoBLockData, err := tx.GetLastEvoBlock()
+	if err != nil {
+		return false, err
+	}
+
+	slog.Debug("IsEvoSyncedWithOwnership", "evo_bock_number", lastEvoBLockData.Number, "evo_block_timestamp", lastEvoBLockData.Timestamp,
+		"lastOwnershipBlock", lastOwnershipBlock, "lastOwnershipBlockTimestamp", lastBlockHeader.Time)
+
+	if lastEvoBLockData.Timestamp < lastBlockHeader.Time {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (p *processor) ProcessUniversalBlockRange(ctx context.Context, startingBlock, lastBlock uint64) error {
+	tx := p.stateService.NewTransaction()
+	defer tx.Discard()
+
+	lastBlockData, err := updateLastBlockData(ctx, tx, p.client, lastBlock)
```

Comment:
ok. yes I agree with you that updating makes more sense to be in the end. It is how it is done in evochain actually. here I had to do at the beggining since I will need last block. but we can put this update in the end.

Reply:
Thanks.
You should also move
```go
slog.Debug("setting ownership end range block hash for block number",
		"blockNumber", block.Number(), "blockHash", block.Hash(), "parentHash", block.ParentHash())
```
right before `tx.SetLastOwnershipBlock`

Reply:
ok